### PR TITLE
[RFC] ci: azure: add job to build and run the Rust tests

### PR DESCRIPTION
**Pull request by jforissier**
Add a job to the Azure CI to build the Rust examples in the
Teaclave Trustzone SDK [1]. They are part of the "buildroot" target so
restrict the build to that particular target to save time.

Link: [1] https://github.com/apache/incubator-teaclave-trustzone-sdk/
Signed-off-by: Jerome Forissier <jerome@forissier.org>

### CODE REVIEW

**Comment by jforissier** [comment]
@fanrong1992 @mssun @DemesneGH please check this out, do you think that would be helpful?
Does the Teaclave project have a test suite that we could run too (similar to the `make check` lines in the `QEMUv8_check` job )?

**Comment by mssun** [comment]
Yes, we have script to run the Rust examples and test with the expected outputs by simply running `cd ci && ./ci.sh` (https://github.com/apache/incubator-teaclave-trustzone-sdk/blob/master/ci/ci.sh).

@fanrong1992 @DemesneGH, can you help to identify how to integrate in the pipeline. Thanks.

**Review by DemesneGH — COMMENTED** [review]
Hi @jforissier 
We already have Github Actions in [there](https://github.com/apache/incubator-teaclave-trustzone-sdk/blob/master/.github/workflows/ci.yml) and I'm working on porting it to Azure Pipelines, like [this](https://github.com/DemesneGH/optee_os/blob/6adc5ed2bc1ba895469d39e580a2b06e482c9be8/azure-pipelines.yml). When I start the job it says `No hosted parallelism has been purchased or granted`. I've sent the request for free parallelism to the Microsoft team and waiting for their reply.

**Comment by mssun** [comment]
@DemesneGH, how about giving some suggestions for this PR to include our test script based on the working version proposed by @jforissier.

**Comment by DemesneGH** [comment]
Hi @mssun 
Since our CI scripts work differently with OP-TEE's, is it more straightforward to write a new test script similar to [qemu-check.exp](https://github.com/OP-TEE/build/blob/master/qemu-check.exp) and add a target `check-rust` in `OP-TEE/build/qemu_v8.mk`?

**Comment by mssun** [comment]
> Hi @mssun
> Since our CI scripts work differently with OP-TEE's, is it more straightforward to write a new test script similar to [qemu-check.exp](https://github.com/OP-TEE/build/blob/master/qemu-check.exp) and add a target `check-rust` in `OP-TEE/build/qemu_v8.mk`?

I agree with you.

@jforissier, how about merging this PR first and we submit changes with `make check-rust` later to the CI job?

**Comment by jforissier** [comment]
@mssun sounds good.

**Comment by jforissier** [comment]
> `Acked-by: Jerome Forissier <jerome@forissier.org>`

Oh, I acked my own pull request :laughing: 

@mssun and/or @DemesneGH would you please give an Acked-by: if you think this is helpful and would like this commit merged? Thanks.

File: .azure-pipelines.yml [hunk 1]
```diff
@@ -242,3 +242,21 @@ jobs:
...
+      - script: |
+          set -e -v
+          export LC_ALL=C
+          WD=$(pwd)
+          sudo -E bash -c "cd /root/optee_repo_qemu_v8/.repo/repo && git pull"
+          sudo -E bash -c "cd /root/optee_repo_qemu_v8 && repo sync -j 10"
+          sudo mv /root/optee_repo_qemu_v8/optee_os /root/optee_repo_qemu_v8/optee_os_old
+          sudo ln -s ${WD} /root/optee_repo_qemu_v8/optee_os
+
+          sudo -E make -C /root/optee_repo_qemu_v8/build -j$(getconf _NPROCESSORS_ONLN) OPTEE_RUST_ENABLE=y CFG_TEE_RAM_VA_SIZE=0x00300000 buildroot
```
Comment (DemesneGH) [inline .azure-pipelines.yml:262]: Maybe we can trigger the `check-rust` progress after `make buildroot`, by:
```
sudo -E make -C /root/optee_repo_qemu_v8/build -j$(getconf _NPROCESSORS_ONLN) CFG_TEE_CORE_LOG_LEVEL=0 OPTEE_RUST_ENABLE=y check-rust
```

Comment (jforissier) [reply .azure-pipelines.yml:262]: OK, I will replace the above line with the one you suggest. It should build & run the thing in one step.

**Comment by DemesneGH** [comment]
`Acked-by: Yuan Zhuang <zhuangyuan04@baidu.com>`

File: .azure-pipelines.yml [hunk 2]
```diff
@@ -242,3 +242,24 @@ jobs:
...
+          export LC_ALL=C
+          WD=$(pwd)
+          sudo -E bash -c "cd /root/optee_repo_qemu_v8/.repo/repo && git pull"
+          sudo -E bash -c "cd /root/optee_repo_qemu_v8 && repo sync -j 10"
+          sudo mv /root/optee_repo_qemu_v8/optee_os /root/optee_repo_qemu_v8/optee_os_old
+          sudo ln -s ${WD} /root/optee_repo_qemu_v8/optee_os
+
+          # FIXME: "make check-rust" should work on its own, but currently it causes "cargo: command not found"
+          sudo -E bash -c "make -C /root/optee_repo_qemu_v8/build -j$(getconf _NPROCESSORS_ONLN) CFG_TEE_CORE_LOG_LEVEL=0 OPTEE_RUST_ENABLE=y optee-rust"
+          sudo -E bash -c "source /root/.cargo/env && make -C /root/optee_repo_qemu_v8/build -j$(getconf _NPROCESSORS_ONLN) CFG_TEE_CORE_LOG_LEVEL=0 OPTEE_RUST_ENABLE=y check-rust"
```
Comment (DemesneGH) [inline .azure-pipelines.yml:265]: Hi @jforissier ,
I found that `source ~/.cargo/bin` in https://github.com/OP-TEE/build/blob/9eccdd56a756d7e669b80c94dabc70fa558c07af/common.mk#L521 doesn't work in some environments.`make OPTEE_RUST_ENABLE=y buildroot` will also cause ` cargo: command not found`.

I've opened a new PR for build/common.mk to fix it:
https://github.com/OP-TEE/build/pull/532

Comment (jforissier) [reply .azure-pipelines.yml:265]: Thanks. With your fix it looks like I can directly invoke the `check-rust` target, at least it worked when I tested with Docker. I will update this PR and hopefully things will go well in Azure too.

**Pull request merged**